modules/src/ics07_tendermint/client_def.rs

```diff
@@ -151,7 +292,73 @@ impl ClientDef for TendermintClient {
         _consensus_state: &Self::ConsensusState,
         _proof_upgrade_client: MerkleProof,
         _proof_upgrade_consensus_state: MerkleProof,
-    ) -> Result<(Self::ClientState, Self::ConsensusState), Error> {
+    ) -> Result<(Self::ClientState, Self::ConsensusState), Ics02Error> {
         todo!()
     }
 }
+
+// Utility method to improve readability of obtaining a Tendermint-specific
+// consensus state from the given context.
+fn read_consensus_state(
+    ctx: &dyn ClientReader,
+    client_id: &ClientId,
+    height: Height,
+) -> Result<ConsensusState, Ics02Error> {
+    maybe_read_consensus_state(ctx, client_id, height)?
+        .ok_or_else(|| Ics02Error::consensus_state_not_found(client_id.clone(), height))
+}
+
+fn maybe_read_consensus_state(
+    ctx: &dyn ClientReader,
+    client_id: &ClientId,
+    height: Height,
+) -> Result<Option<ConsensusState>, Ics02Error> {
+    match ctx.consensus_state(client_id, height).map(|cs| {
+        downcast!(
+            cs => AnyConsensusState::Tendermint
+        )
+        .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))
+    }) {
+        Ok(result) => result.map(Some),
+        Err(e) => match e.detail() {
+            Ics02ErrorDetail::ConsensusStateNotFound(_) => Ok(None),
+            _ => Err(e),
+        },
+    }
+}
+
+// Searches forwards for updates where lower_height < h < upper_height
+fn maybe_read_next_consensus_state(
+    ctx: &dyn ClientReader,
+    client_id: &ClientId,
+    lower_height: Height,
+    upper_height: Height,
```

Comment:
Wondering here about performance, as we incrementally look for a state with height +1. The heights of the consensus states are like swiss cheese, i.e. there could be a big gap between lower and upper heights. An alternative is to have `maybe_read_next_consensus_state()` (no need for `upper_height`) added to the `ClientReader` and let each context implement it. For the mock it could do a linear search through the consensus state keys, others may implement a more efficient search structure.
```suggestion
fn maybe_read_next_consensus_state(
    ctx: &dyn ClientReader,
    client_id: &ClientId,
    height: Height,
```